roofit/roofitcore/res/RooFit/ADModeHelpers.h

```diff
@@ -0,0 +1,89 @@
+/*
+ * Project: RooFit
+ * Authors:
+ *   Garima Singh, CERN 2022
+ *
+ * Copyright (c) 2022, CERN
+ *
+ * Redistribution and use in source and binary forms,
+ * with or without modification, are permitted according to the terms
+ * listed in LICENSE (http://roofit.sourceforge.net/license.txt)
+ */
+
+#ifndef RooFit_ADModeHelpers_h
+#define RooFit_ADModeHelpers_h
+
+#include <RooGlobalFunc.h>
+#include <RooAbsReal.h>
+#include <RooListProxy.h>
+
+#include <memory>
+#include <string>
+
+class RooAbsData;
+class RooAbsPdf;
+class RooAbsReal;
+class RooArgSet;
+
+/// @brief  A wrapper class to store the generated function of a given RooFit workspace.
+/// @tparam Func Function pointer to the generated function.
+template <typename Func = void>
+class RooGradFuncWrapper final : public RooAbsReal {
+public:
+   RooGradFuncWrapper(Func function, std::string funcName, RooArgSet const &paramSet)
+      : RooAbsReal{"RooGradFuncWrapper", "RooGradFuncWrapper"}, _paramProxies{"!params", "List of parameters", this},
+        _funcName(funcName), _func(function)
+   {
+      for (auto *param : paramSet) {
+         _paramProxies.add(*param);
+         if (auto realParam = dynamic_cast<RooRealVar *>(param))
+            _params.emplace_back(realParam);
+      }
+   }
+
+   RooGradFuncWrapper(const RooGradFuncWrapper &other, const char *name = nullptr)
+      : RooAbsReal(other, name), _paramProxies("!params", this, other._paramProxies), _func(other._func)
+   {
+   }
+
+   TObject *clone(const char *newname) const override { return new RooGradFuncWrapper(*this, newname); }
+
+   double defaultErrorLevel() const override { return 0.5; }
+
+protected:
+   double evaluate() const override
+   {
+      std::vector<double> paramVals(_paramProxies.size());
```

Comment:
```suggestion
      std::vector<double> paramVals;
      paramVals.reserve(_paramProxies.size());
```
Otherwise you have a vector with size `2 * npars`, where the first half is just filled with zeros!